# Using functional connectivity from the HCP to predict individual characteristics

Connectivity ML Group @ [Neurohackademy 2021](https://neurohackademy.org/)

## But first, import!

In [1]:
# our core libraries
import math
import numpy as np
import pandas as pd
import neuropythy as ny
import nibabel as nib
import ipyvolume as ipv
import random
import matplotlib.pyplot as plt
import seaborn as sns

from tqdm import tqdm

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report,plot_confusion_matrix

from sklearn.pipeline import make_pipeline

from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.linear_model import SGDRegressor, LogisticRegression
from sklearn.ensemble import RandomForestClassifier

import shap  # package used to calculate Shap values

In [3]:
# You need to configure neuropythy so that it knows what your
# HCP AWS S3 access key and secret are:
ny.config['hcp_credentials'] = (key, secret)

ny.config['hcp_auto_download'] = True
ny.config['hcp_auto_path'] = '~/hcp_data'

# Next, load the data..

To `netmaps_df` we load "netmaps" which are subject-specific “parcellated connectomes” – for each subject, a nodes x nodes network matrix. See more [here](https://www.humanconnectome.org/storage/app/media/documentation/s1200/HCP1200-DenseConnectome+PTN+Appendix-July2017.pdf).

To `behavioral_df` we load the data keys. See more [here](https://wiki.humanconnectome.org/display/PublicData/HCP-YA+Data+Dictionary-+Updated+for+the+1200+Subject+Release). 

In [4]:
N = 15 # number of ICAs - 15, 25, 50 ,100 ,200 , 300

In [5]:
netmaps_df = pd.read_csv('data/connectivityml/HCP_PTN1200/netmats/3T_HCP1200_MSMAll_d'+str(N)+'_ts2/netmats2.txt', delim_whitespace=True,header=None)
print("Network-matrices data shape:", netmaps_df.shape)
netmaps_df.head()

Network-matrices data shape: (1003, 225)


0        1       2        3       4       5        6        7        8    \
0    0 -6.43160  2.9077   5.6367 -1.9108 -2.9030 -0.78089   1.6558 -0.46461   
1    0 -4.09320 -4.8082   4.2359  1.2897 -6.0519  0.11149  -6.8054 -2.10540   
2    0 -0.15331 -4.4759  11.8270 -4.2061 -6.5727 -1.61310 -10.2110 -1.28100   
3    0 -0.87383 -6.8278  10.1640 -1.8393 -9.1321 -2.86400  -7.2182  1.55410   
4    0 -3.96500  2.5420  11.5050 -4.9034 -2.4797  3.18270  -4.2678  1.84570   

      9    ...     215      216      217     218     219     220     221  \
0  5.3488  ...  3.4614  0.89082 -4.27550  5.1550 -6.4641 -5.1331  5.8573   
1  2.7563  ...  6.3430  0.35134 -0.67089  7.6578 -3.8168 -3.7186  1.4174   
2  2.0158  ...  2.4110  3.49980 -1.95180  2.4394 -9.2378 -3.5996  4.5837   
3  2.2322  ... -3.2454  0.34570 -1.92850  6.4472 -6.9826 -1.2231  6.6318   
4  5.9304  ...  2.2238  7.66490  1.33870  8.1098 -1.4201 -1.3990  8.1015   

        222     223  224  
0 -2.723100  6.1086    0  
1 -2.052100  5.0104    0  
2 -3.172500  4.0976    0  
3 -0.667030  7.3691    0  
4  0.014264  1.4994    0  

[5 rows x 225 columns]

In [ ]:
behavioral_df = pd.read_csv('data/connectivityml/unrestricted_pkalra_7_26_2021_17_39_25.csv')
print("Behaviora data shape:", behavioral_df.shape)
behavioral_df.head()

We have netmaps for 1003 subjects so we will need to filter `behavioral_df` a little.

To `subjectsID_df` we load the ordered list of all subjects with complete rfMRI data (recon 1 + recon2) included in this PTN release

In [ ]:
subjectsID_df = pd.read_csv('data/connectivityml/HCP_PTN1200/subjectIDs.txt',header=None,names=["Subject"])
print("Subjects ID data shape:", subjectsID_df.shape)
subjectsID_df.head()

We can see that this corresponds to the number of netmaps we have.

In [ ]:
filter_behavioral_df = subjectsID_df.merge(behavioral_df, on='Subject', how='inner')

print("Filtered behaviora data shape:", filter_behavioral_df.shape)
filter_behavioral_df.head()

## Pre-process features matrix

In [ ]:
netmapsX_df = pd.DataFrame(data = netmaps_df, columns = range(N*N))
netmapsX_df = netmapsX_df.T.drop_duplicates(keep='first').T
netmapsX_df = netmapsX_df.T.drop_duplicates(keep='last').T
X = netmapsX_df
print("Features matrix shape:", X.shape)
X.head()

## Pre-process predicted values

Here we are going to foucs on the subject gender.

In [ ]:
filter_behavioral_df['Gender_i']=np.zeros(shape=(subjectsID_df.shape))
filter_behavioral_df.Gender_i = pd.factorize(filter_behavioral_df.Gender)[0] # Encode the object as an enumerated type or categorical variable.
y_gender = filter_behavioral_df.Gender_i # Gender of Subject
print("y_gender shape:", y_gender.shape)

filter_behavioral_df['Gender_i'].groupby(filter_behavioral_df['Gender']).unique().apply(pd.Series).rename(columns={0:'Labels'}).sort_values(by='Labels')

In [ ]:
fig, ax = plt.subplots()
sns.histplot(y_gender, ax=ax)
ax.set_title("Gender of Subject")
fig.tight_layout()

## Time for Random forests!

Data and estimators exploration was done in a separate notebook..

### [RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

A random forest is a meta estimator that fits a number of decision tree classifiers on various sub-samples of the dataset and uses averaging to improve the predictive accuracy and control over-fitting.

Here we will prdict the gender of subjects in our dataset from our network-matrices data has gave us the best results...

In [ ]:
xtrain, xtest, ytrain, ytest = train_test_split(X, y_gender, test_size=0.2,stratify=y_gender,random_state=1)

In [ ]:
rf = RandomForestClassifier(random_state=1)

rf.fit(xtrain, ytrain) # Build a forest of trees from the training set (X, y).

score = rf.score(xtest, ytest) # Return the mean accuracy on the given test data and labels.
print("Test set score: ", score) 

ypred = rf.predict(xtest) # Predict class for X.

In [ ]:
# View confusion matrix for test data and predictions
plot_confusion_matrix(rf, xtest, ytest) # Plot Confusion Matrix.
plt.savefig('output/confusion_matrix_'+str(N)+'.png')

In [ ]:
cr = classification_report(ytest, ypred,output_dict=True) # Build a text report showing the main classification metrics.
sns_plot = sns.heatmap(pd.DataFrame(cr).iloc[:-1, :].T, annot=True) # plot scikit-learn classification report
sns_plot.figure.savefig("output/classification_report_"+str(N)+".png")

#### Visualizing trees
We can plot individual decision trees - 

In [ ]:
from sklearn.tree import export_graphviz
from subprocess import call
from IPython.display import Image, display

def plot_graphviz_tree(tree):
    """
    Helper function that takes a tree as input, calls sklearn's export_graphviz
    function to generate an image of the tree using graphviz, and then
    plots the result in-line.
    """
    export_graphviz(tree, out_file='tree.dot', max_depth=3, filled=True,
                    feature_names=X.columns, impurity=False, rounded=True,
                    proportion=False, precision=2);

    call(['dot', '-Tpng', 'tree.dot', '-o', 'output/tree_'+str(N)+'.png', '-Gdpi=600'])
    display(Image(filename = 'output/tree_'+str(N)+'.png'));

In [ ]:
# First tree in the forest
plot_graphviz_tree(rf.estimators_[0]);

#### Interpreting random forests

##### Feature importances

Unlike regression-based methods, random forests don't have linear coefficientsso we look at the feature importances, to tell us how each feature contributes to the overall prediction.

In [ ]:
pd.Series(rf.feature_importances_, index=X.columns).sort_values(ascending=False).head(10)

##### SHAP values
To interpret our results we will look at the SHAP values of our model.
See more [here](https://towardsdatascience.com/explain-any-models-with-the-shap-values-use-the-kernelexplainer-79de9464897a).

SHAP feature importance is an alternative to standard feature importance based on magnitude of feature attributions. The feature importance is useful, but contains no information beyond the importances. For a more informative plot, we will next look at the summary plot.

The SHAP summary plot is made of all the dots in the test data. Showing:
- **Feature importance:** Variables are ranked in descending order.
- **Impact:** The horizontal location shows whether the effect of that value is associated with a higher or lower prediction.
- **Original value:** Color shows whether that variable is high (in red) or low (in blue) for that observation.


In [ ]:
# Create object that can calculate shap values
explainer = shap.TreeExplainer(rf)

# calculate shap values. This is what we will plot.
# Calculate shap_values for all of xtest rather than a single row, to have more data for plot.
shap_values = explainer.shap_values(xtest)

# Make plot. Index of [1] is explained in text below.

fig = shap.summary_plot(shap_values[1], xtest, show=False)
plt.savefig('output/summary_plot_xtest_'+str(N)+'.png')

#### Single subject interpretation
We will look at SHAP values for a random row of the dataset.

##### SHAP values
For context, we'll look at the raw predictions before looking at the SHAP values.

In [ ]:
row_to_show = 5
data_for_prediction = xtest.loc[row_to_show]  # use 1 row of data here. Could use multiple rows if desired
data_for_prediction_array = data_for_prediction.values.reshape(1, -1)

rf.predict_proba(data_for_prediction_array)

In [ ]:
print("Subject evaluated:",subjectsID_df.Subject[row_to_show])

In [ ]:
# Create object that can calculate shap values
explainer = shap.TreeExplainer(rf)

# Calculate Shap values
shap_values = explainer.shap_values(data_for_prediction)
shap.initjs()

shap.force_plot(explainer.expected_value[1], shap_values[1], data_for_prediction, show=False)
# plt.savefig('force_plot_hcp_'+str(subjectsID_df.Subject[row_to_show])+'_'+str(N)+'.png')

The SHAP values of all features sum up to explain why the prediction was different from the baseline. This allows us to decompose a prediction in a graph like this where:

- The **output value** is the prediction for that observation (the prediction of the subject evluated).
 - The **base value** is the value that would be predicted if we did not know any features for the current output (the mean prediction, or mean(yhat)).
 - **Red/blue:** Features that push the prediction higher (to the right) are shown in red, and those pushing the prediction lower are in blue.
 
##### TOP 5 Most important ICA's

In [ ]:
vals= np.abs(shap_values).mean(0)
feature_importance = pd.DataFrame(list(zip(xtest.columns,vals)),columns=['netmat_col','feature_importance_vals'])
feature_importance.sort_values(by=['feature_importance_vals'],ascending=False,inplace=True)
feature_importance.head()

In [ ]:
edge_df = pd.DataFrame(columns = ['netmat_col','Node1', 'Node2', 'Weight'])
countlist = list() # if node1,node2 weight is saved no need to save node2,node1

for index, row in tqdm(netmaps_df.iterrows(), total=netmaps_df.shape[0]):
    netmat_col = 0
    if index == row_to_show:
        row2mat = row.values.reshape(N,N)
        for node1 in range(N):
            for node2 in range(N):
                if node1!=node2:
                    if (node2, node1) not in countlist:
                        countlist.append((node1, node2))
                        curr_edge = {'netmat_col': netmat_col, 'Node1': node1, 'Node2': node2, 'Weight':row2mat[node1][node2]}
                        edge_df = edge_df.append(curr_edge, ignore_index = True)
                netmat_col = netmat_col + 1

In [ ]:
f = open('output/top5_'+str(subjectsID_df.Subject[row_to_show])+'_'+str(N)+'.txt', "w")

for i in range(5):
    feature = feature_importance.netmat_col.values[i]
    print("Important feature #: ",feature)
    f.write("Important feature #: " + str(feature)+"\n")
    ICA_Node1 = edge_df.loc[edge_df['netmat_col'] == feature, 'Node1']
    ICA_Node2 = edge_df.loc[edge_df['netmat_col'] == feature, 'Node2']

    print("Node 1:",ICA_Node1.iloc[0])
    f.write("Node 1: " + str(ICA_Node1.iloc[0])+"\n")
    print("Node 2:",ICA_Node2.iloc[0])
    f.write("Node 2: " + str(ICA_Node2.iloc[0])+"\n")
f.close()

Knowing the ICA's that are part of the most important feature we can plot them on our single subject brain

In [ ]:
# Get a sample HCP subject:
sub = ny.hcp_subject(subjectsID_df.Subject[row_to_show])
sub

In [ ]:
# Load the CIFTI file:
cii_filename = '~/data/connectivityml/HCP_PTN1200/groupICA/groupICA_3T_HCP1200_MSMAll_d'+str(N)+'.ica/melodic_IC.dscalar.nii'
cii_obj = ny.load(cii_filename)

# Split the CIFTI object into hemisphere/subvoxel data:
(lh_data, rh_data, subvox_data) = ny.hcp.cifti_split(cii_obj)

# These data should be (N(data-points) x vertices)
lh_data.shape

In [ ]:
# sub.lh and sub.rh are the "native" (FreeSurfer) hemispheres;
# sub.hemis['lh_LR32k'] and sub.hemis['rh_LR32k'] are the
# HCP subject-aligned fs_LR hemispheres (with 32k resolution).
lh_hemi_native = sub.lh
rh_hemi_native = sub.rh

# The 32492 size indicates this is a 32k LR hemisphere:
lh_hemi = sub.hemis['lh_LR32k']
rh_hemi = sub.hemis['rh_LR32k']

**ICA 1**

In [ ]:
# We can make an ipyvolume figure to plot both hemispheres on:
fig = ipv.figure()
# Then plot each hemisphere using whichever ICA component we
# want to visualize:

ICA_Node1 = int(edge_df.loc[edge_df['netmat_col'] == feature_importance.netmat_col.values[0], 'Node1'])
ny.cortex_plot(lh_hemi, surface='inflated', color=lh_data[ICA_Node1], cmap='hot', figure=fig)
ny.cortex_plot(rh_hemi, surface='inflated', color=rh_data[ICA_Node1], cmap='hot', figure=fig)

In [ ]:
ipv.pylab.save('output/most_important_ICA1_'+str(subjectsID_df.Subject[row_to_show])+'_'+str(N)+'.html',title='Most important feature ICA 1')

**ICA 2**

In [ ]:
# We can make an ipyvolume figure to plot both hemispheres on:
fig = ipv.figure()
# Then plot each hemisphere using whichever ICA component we
# want to visualize:

ICA_Node2 = int(edge_df.loc[edge_df['netmat_col'] == feature_importance.netmat_col.values[0], 'Node2'])
ny.cortex_plot(lh_hemi, surface='inflated', color=lh_data[ICA_Node2], cmap='hot', figure=fig)
ny.cortex_plot(rh_hemi, surface='inflated', color=rh_data[ICA_Node2], cmap='hot', figure=fig)

In [ ]:
ipv.pylab.save('output/most_important_ICA2_'+str(subjectsID_df.Subject[row_to_show])+'_'+str(N)+'.html',title='Most important feature ICA 2')